### ISSUE TITLE
Output SPIR-V is invalid: ID 22[%22] defined in block 11[%11] does not dominate its use in block 24[%24]

### ISSUE THREAD

**Issue by paulthomson** [KhronosGroup/glslang#1822]
[bug_report.zip](https://github.com/KhronosGroup/glslang/files/3355488/bug_report.zip)

Issue found using [GraphicsFuzz](https://github.com/google/graphicsfuzz).

Tool versions:

* glslangValidator commit hash: 40c16ec0b3ad03fc170f1369a58e7bbe662d82cd

To reproduce:

`glslangValidator -V shader.comp -o shader.comp.spv`

`spirv-val shader.comp.spv`

The following shader files are included in the attached archive, some of which are also shown inline below:

* 0_glsl/shader.comp
* 1_spirv/shader.comp.spv
* 1_spirv_asm/shader.comp.asm

0_glsl/shader.comp:

```
#version 310 es

layout(set = 0, binding = 1) uniform buf1 {
 vec2 injectionSwitch;
};
layout(std430, binding = 0) buffer theSSBO {
 bool is_exclusive;
 uint data_out[200];
 uint data_in[200];
} ;
layout(local_size_x = 100, local_size_y = 1, local_size_z = 1) in;
void main()
{
 if((true ? injectionSwitch.x : fwidth(17.39)) > 1.0)
  {
  }
}

```

1_spirv_asm/shader.comp.asm:

```
; SPIR-V
; Version: 1.0
; Generator: Khronos Glslang Reference Front End; 7
; Bound: 41
; Schema: 0
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %4 "main"
               OpExecutionMode %4 LocalSize 100 1 1
               OpSource ESSL 310
               OpName %4 "main"
               OpName %14 "buf1"
               OpMemberName %14 0 "injectionSwitch"
               OpName %16 ""
               OpName %34 "theSSBO"
               OpMemberName %34 0 "is_exclusive"
               OpMemberName %34 1 "data_out"
               OpMemberName %34 2 "data_in"
               OpName %36 ""
               OpMemberDecorate %14 0 Offset 0
               OpDecorate %14 Block
               OpDecorate %16 DescriptorSet 0
               OpDecorate %16 Binding 1
               OpDecorate %32 ArrayStride 4
               OpDecorate %33 ArrayStride 4
               OpMemberDecorate %34 0 Offset 0
               OpMemberDecorate %34 1 Offset 4
               OpMemberDecorate %34 2 Offset 804
               OpDecorate %34 BufferBlock
               OpDecorate %36 DescriptorSet 0
               OpDecorate %36 Binding 0
               OpDecorate %40 BuiltIn WorkgroupSize
          %2 = OpTypeVoid
          %3 = OpTypeFunction %2
          %6 = OpTypeBool
          %7 = OpConstantTrue %6
          %8 = OpTypeFloat 32
          %9 = OpTypePointer Function %8
         %13 = OpTypeVector %8 2
         %14 = OpTypeStruct %13
         %15 = OpTypePointer Uniform %14
         %16 = OpVariable %15 Uniform
         %17 = OpTypeInt 32 1
         %18 = OpConstant %17 0
         %19 = OpTypeInt 32 0
         %20 = OpConstant %19 0
         %21 = OpTypePointer Uniform %8
         %27 = OpConstant %8 1
         %31 = OpConstant %19 200
         %32 = OpTypeArray %19 %31
         %33 = OpTypeArray %19 %31
         %34 = OpTypeStruct %19 %32 %33
         %35 = OpTypePointer Uniform %34
         %36 = OpVariable %35 Uniform
         %37 = OpTypeVector %19 3
         %38 = OpConstant %19 100
         %39 = OpConstant %19 1
         %40 = OpConstantComposite %37 %38 %39 %39
          %4 = OpFunction %2 None %3
          %5 = OpLabel
         %10 = OpVariable %9 Function
               OpSelectionMerge %12 None
               OpBranchConditional %7 %11 %24
         %11 = OpLabel
         %22 = OpAccessChain %21 %16 %18 %20
         %23 = OpLoad %8 %22
               OpStore %10 %23
               OpBranch %12
         %24 = OpLabel
         %25 = OpLoad %8 %22
               OpStore %10 %25
               OpBranch %12
         %12 = OpLabel
         %26 = OpLoad %8 %10
         %28 = OpFOrdGreaterThan %6 %26 %27
               OpSelectionMerge %30 None
               OpBranchConditional %28 %29 %30
         %29 = OpLabel
               OpBranch %30
         %30 = OpLabel
               OpReturn
               OpFunctionEnd

```

**Comment by afd** [comment]
Like #1823, the input GLSL is invalid due to fwidth only being available in fragment shaders (we had a bug in GraphicsFuzz where we would introduce such builtins to compute shaders).

@johnkslang worth validating that compute shaders do not use fragment processing builtins?

**Comment by johnkslang** [comment]
Nvidia has recent extensions that add these fragment-only built-ins to the compute stage.  I suspect it was done too broadly.

**Comment by johnkslang** [comment]
BTW, I have a side project going that simplifies how this is done internally to glslang, considering the current way as technical debt worth fixing.

**Comment by arcady-lunarg** [comment]
This was fixed by #1837. Closing.
